### ISSUE TITLE
Develop Diffusive-Reservoir integration

### ISSUE THREAD

**Issue by jameshalgren** [NOAA-OWP/t-route#332]
@jdmattern-noaa, @mehdire61 -- these first three items are what we discussed today on the call (we identified only two, explicitly, but the first is implicitly required to achieve the second two.)
### Create a new parallel option in compute.py, mimicking the `by-subnetwork-jit` 
Essentially, [these lines](https://github.com/NOAA-OWP/t-route/blob/master/src/python_routing_v02/troute/routing/compute.py#L333-L552) provide a template for a new option which could be named `by-subnetwork-diffusive`. For the moment, there will be only one option for subnetwork splitting (that is, on the reservoirs) but we can explore the idea of having another piece of logic that allows additional subnetwork splitting schemes. 

### Add a function for subnetwork development by reservoirs; add to new parallel option
@awlostowski-noaa, will you please add a reference to the prototype code you developed?
The new function would replace [this call](https://github.com/NOAA-OWP/t-route/blob/master/src/python_routing_v02/troute/routing/compute.py#L334-L336) in the new parallel option.

### Add logic for switching compute kernels depending on content of subnetwork
Some additional logic, ultimately affecting `compute_func` as used in [this call](https://github.com/NOAA-OWP/t-route/blob/master/src/python_routing_v02/troute/routing/compute.py#L496), will be needed to make sure reservoir subnetworks (which will really just be single segments) are sent to the compute kernel for reservoirs while routing subnetworks are sent to the diffusive routing kernel. Eventually, this switching capability could provide the basis for a totally heterogenous routing strategy (assuming we can find a way to satisfy the physics requirements in the subnetwork  

@kumdonoaa @awlostowski-noaa -- this piece is where we may need to do some work on the diffusive kernel
### Add handling of upstream inputs for diffusive simulation to support reservoir+diffusive integration
Somewhere as part of [this call](https://github.com/NOAA-OWP/t-route/blob/82e7e62d74ac1ec8e876932921cb42adf31006e4/src/python_routing_v02/troute/routing/fast_reach/diffusive.pyx#L144), we need to add code to prepare a dataset of upstream inflows that will be coming as part of the integration with subnetwork-based reservoir modeling. The MC algorithm does this by adding the flows to the `flowveldepth` array; some other method may be more appropriate for the diffusive algorithm.

**Comment by awlostowski-noaa** [comment]
Here is the subnetwork creation prototype https://github.com/awlostowski-noaa/t-route/blob/break-nets-at-reservoirs/notebooks/break-networks-at-reservoirs.ipynb

**Comment by jameshalgren** [comment]
Follow-on work possibilities: 
- [ ] expand from Croton to Florence for testing
- [ ] Evaluate performance of splitting algorithm on large-scale domain (i.e., CONUS)
- [ ] Complete merge with 'buttoned-up' test case
- [ ] scope (and possibly begin) follow-on effort with more sophisticated splitting scheme 
Ideas:
* determine need (does reservoir-only satisfy the operational requirement? can we be strategic about clustering to satisfy the parallel execution need?)
* document split network statistics (make tools to do this and report on it as part of the nhd_network script/network package) (how many networks are we seeing? What are their sizes and size distributions? What are the topological relationships, i.e., what's the longest path?)
* build simple 'overlap' test (for first proposed method, we see the idea of an 'overlap' zone; the mechanics and logic of handling such a zone need to be added to the code. It could be useful to generalize this capability to be functional not only in the diffusive subnetwork parallel mode.)
* Build a markdown writeup explaining the proposed method
* build and evaluate network size from 'overlap' splitting schemes
* build and evaluate tests results generated from applying other splitting schemes

**Comment by awlostowski-noaa** [comment]
# Thoughts on passing reservoir outflows to diffusive wave model as upstream boundary conditons
- The upstream boundary condition data for the diffusive wave model is contained in an array named `ubcd_g`
- The `ubcd_g` array is created by a function named [`fp_ubcd_map` in `diffusive_utils.py`](https://github.com/NOAA-OWP/t-route/blob/82e7e62d74ac1ec8e876932921cb42adf31006e4/src/python_routing_v02/troute/routing/diffusive_utils.py#L296-L326)
    - This function identifies headwater segments and writes a series of boundary condition flows for these segments
    - currently, lateral inflow data are written-in as upstream boundary condition data, like this:
    ```(python) ubcd_g[tsi, frj] = qlat_data[idx_segID, tsi]  # [m^3/s]  ```
- I suggest we edit this function such that flows from off-network upstream segments are written in, instead of lateral inflows. 
    - specifically, for the case when `idx_segID` is in the collection of off-network upstream segments (which should be a group of reservoirs)

**Comment by awlostowski-noaa** [comment]
Here is an example of how we would extract the upstream flow values from the `upstream_results` dictionary:

``` python
    for upstream_tw_id in upstream_results:
        tmp = upstream_results[upstream_tw_id]
        fill_index = tmp["position_index"]
        fill_index_mask[fill_index] = False
        for idx, val in enumerate(tmp["results"], qvd_ts_w):
            flowveldepth[fill_index, idx] = val
```

**Comment by awlostowski-noaa** [comment]
Need to change the dimensions ubcd_g in Fortran code to allow finer time resolution of reservoir outflow data

**Comment by jameshalgren** [comment]
Closed by #347
